fix #282: datetime column in associate table based on connection dialect

indiVar0508 opened the pull request:
attempt to address #282, to process the input parameters received in @track_association_operations method

before changes Pytest results
346 failed, 308 passed, 14 skipped, 5398 warnings, 90 errors
After Changes Pytest results
346 failed, 308 passed, 14 skipped, 5398 warnings, 90 errors

indiVar0508 commented:
Hi @marksteward ,

Did you get a chance to review this PR ?

marksteward commented:
Not yet.

AbdealiJK commented on sqlalchemy_continuum/manager.py:
```diff
@@ -431,6 +431,21 @@ def track_cloned_connections(self, c, opt):
+            return process(parameter)
+        return parameter
+
+    def process_parameters(self, table_name, parameters, dialect):
```
(Opinionated) We don't really need 2 functions here - we can do it in a single function with a for loop ?

indiVar0508 replied:
yeah sure we can do that, kept them separate as method `get_python_processed_parameter` uses dialect to process parameters so if any dialect specific condn (process being None, not aware of any other cond's as of now) is identified the handling can be dealt within that method itself

AbdealiJK commented on sqlalchemy_continuum/manager.py:
```diff
@@ -431,6 +431,21 @@ def track_cloned_connections(self, c, opt):
+        table = self.metadata.tables[table_name]
+        return tuple(
+            self.get_python_processed_parameter(parameter=parameters[idx], column=column, dialect=dialect) 
+            for idx, column in enumerate(table.columns)
```
I am not sure I get this.
You are doing a `enumerate(table.columns)` - so lets say I have 10 columns in my table.
Not you do `parameters[idx]` - You are assuming every parameter corresponds to each column ?

There are some asusmptions you seem to be making:
1. This is only valid for INSERT statements
2. Are we sure that ALL columns will be present ? What if there is a `server_default` or nullabe=True which makes the column null by default ?

AbdealiJK replied:
In fact, now that I look at it more - isn't `parameters` a dictionary ?
Shouldn't we be doing: `parameters[column.key]` ?

indiVar0508 replied:
yes, i think for defaults it does return some value from SQLA itself, however the issue you are pointing to i have raised same in thread for #283 for delete query perspective and i think it does exist currently in master branch also, a seperate thread is on ig for refactoring `positional_args_to_dict`

` isn't parameters a dictionary ?` - No it's a tuple `positional_args_to_dict` converts this tuple to dict

AbdealiJK commented on sqlalchemy_continuum/manager.py:
```diff
@@ -431,6 +431,21 @@ def track_cloned_connections(self, c, opt):
+        param_transformer = column.type.dialect_impl(dialect)
+        process = param_transformer.result_processor(dialect, column.type)
+        if process is not None:
+            # https://github.com/sqlalchemy/sqlalchemy/blob/231262d4782fc4d1a5d5fdbf985a06e04270e878/lib/sqlalchemy/sql/type_api.py#L452
```
When adding a link unless you explicitly need this commit - you could just use the last stable version:
https://github.com/sqlalchemy/sqlalchemy/blob/rel_1_4_40/lib/sqlalchemy/sql/type_api.py#L335
THe link becomes easier to read also.
Plus you're linking to the function there - it would be ideal to explain what you want us to read there.
For example:
```
If `result_processor` is None - no processing is required as per sqlalchemy
```

indiVar0508 replied:
Sure will update this one and keep that in check for future PR(s)

indiVar0508 commented:
Hi @AbdealiJK @marksteward ,

Can you review changes for [address_postional_agrs issue](https://github.com/kvesteri/sqlalchemy-continuum/pull/295/commits/1ee5c41656a94d475ff9142ef15d6b9509c5d581),

i was looking into this and i think new chnages address #280 and #283 / #284, however while deleting entries i am getting #275 not sure why it seems for delete command `units_of_work` is not being populated in before_flush or before_flush is not being invoked for delete commands

pytest is working similar to latest master branch
`345 failed, 311 passed, 14 skipped, 302 warnings, 90 errors`

i was trying with below mentioned cases
```
# #280
# lotr = Book(name='Lord of the rings')
# tolkien = Author(name='JRR Tolkien', books=[lotr])
# db.add(lotr)
# db.add(tolkien)
# db.commit()
# #284/#283(PR thread)
from sqlalchemy import and_
d = book_author_table.delete(whereclause=and_(book_author_table.c.author_id==1, book_author_table.c.book_id==1))
# d = book_author_table.delete()
print(d)
engine.execute(d)
db.commit()
```
but when delete operation is triggered the statement get processed and params also seems to be correct 
`{'author_id_1': 1, 'book_id_1': 1}`

but i get a error with below mentioned traceback

```
Traceback (most recent call last):
  File "/home/indivar/github/sqlalchemy-continuum/sqlalchemy_continuum/manager.py", line 411, in append_association_operation
    uow = self.units_of_work[conn]
KeyError: <sqlalchemy.engine.base.Connection object at 0x7f5abadd7df0>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "distdemo_books.py", line 66, in <module>
    engine.execute(d)
  File "<string>", line 2, in execute
  File "/home/indivar/github/py_envs/sqlalchemy_code_update_venv/lib/python3.8/site-packages/sqlalchemy/util/deprecations.py", line 402, in warned
    return fn(*args, **kwargs)
  File "/home/indivar/github/py_envs/sqlalchemy_code_update_venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 3257, in execute
    return connection.execute(statement, *multiparams, **params)
  File "/home/indivar/github/py_envs/sqlalchemy_code_update_venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1380, in execute
    return meth(self, multiparams, params, _EMPTY_EXECUTION_OPTS)
  File "/home/indivar/github/py_envs/sqlalchemy_code_update_venv/lib/python3.8/site-packages/sqlalchemy/sql/elements.py", line 333, in _execute_on_connection
    return connection._execute_clauseelement(
  File "/home/indivar/github/py_envs/sqlalchemy_code_update_venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1572, in _execute_clauseelement
    ret = self._execute_context(
  File "/home/indivar/github/py_envs/sqlalchemy_code_update_venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1842, in _execute_context
    statement, parameters = fn(
  File "/home/indivar/github/py_envs/sqlalchemy_code_update_venv/lib/python3.8/site-packages/sqlalchemy/engine/events.py", line 128, in wrap_before_cursor_execute
    orig_fn(
  File "/home/indivar/github/sqlalchemy-continuum/sqlalchemy_continuum/manager.py", line 475, in track_association_operations
    self.append_association_operation(
  File "/home/indivar/github/sqlalchemy-continuum/sqlalchemy_continuum/manager.py", line 414, in append_association_operation
    uow = self.units_of_work[conn.engine]
KeyError: Engine(sqlite:///db.sqlite)
```

indiVar0508 commented:
Hi @marksteward , 

I have tidied the commits for this PR and tested the issue #282 for SQLA < 1.4, 1.3 and 1.2 it seems to works , could you review it and run travis CI Build

Pytest Run results : `346 failed, 308 passed, 14 skipped, 5398 warnings, 90 errors`

indiVar0508 commented:
It seems it is failing for postgres-native one, if you get a chance, could you also provide  traceback for testcase so i will try to debug the issue locally.

indiVar0508 commented:
Hi @marksteward ,

I have fixed the issue for last action [job run](https://github.com/kvesteri/sqlalchemy-continuum/runs/8117534892?check_suite_focus=true), the job run was failing becuase in association i missed schema name consideration for table for SQLA > 1.4, i have added that now and tested with postgres in local for failing testfile(tests/relationships/test_many_to_many_relations.py) and it seems to work now.

Could you again trigger the Travis Build?

indiVar0508 commented:
Hi @marksteward ,

The Travis Build seems to be success for this one, is there any other changes that are required for this PR or can accept the request and close it ?

this addresses #282 and #284

AbdealiJK commented:
Hi @indiVar0508 thanks for working on this.
I was not aware of `context.compiled_parameters` - it's awesome that dwlalchemy already provides this instead of us doing all the processing we were doing earlier.

Just 2 comments:
1. Can you squash your commits as they both are solving the same thing
2. Can you add a testcase for. The scenario where an association-table has a date field ? This was we have a testcase for the reported bug.

@marksteward can this be tagged as `next-release` too ? The approach seems to be clear and this is an issue for my application

indiVar0508 commented:
Hi @AbdealiJK ,

Updated PR with mentioned changes

marksteward commented:
Hi @indiVar0508, @AbdealiJK. Yes, I love the approach, particularly getting rid of `positional_args_to_dict`, but I want to make sure it covers any weird corner cases. Happy to add it to the next release.

indiVar0508 commented:
Hi @marksteward, 

sure if you want i can try to look into some corner cases and test it, can you give some inputs to point me the direction on what cases you want me to check i can try them locally
